### PR TITLE
[Heng Raft]Heartbeat

### PR DESCRIPTION
**Pull request by flyingmeat**

### CODE REVIEW

File: lab/src/raft/leader_election.go [hunk 1]
```diff
@@ -1,11 +1,32 @@
...
+	// Send heartbeat to all other followers
+	for server := 0; server < len(rf.peers); server++ {
+		if server == rf.me {
+			continue
+		}
+		rf.mu.Lock()
+		currentTerm := rf.currentTerm
+		rf.mu.Unlock()
+		go func(currentTerm int, server int) {
+			args := AppendEntriesArgs{currentTerm, rf.me, []*LogEntry{}, 0, 0, rf.committedIndex}
```
Comment (liana-yang) [inline lab/src/raft/leader_election.go:19]: Add `LeaderId` to AppendEntriesArgs.

File: lab/src/raft/append_entries.go [hunk 2]
```diff
@@ -40,7 +40,9 @@ func (rf *Raft) AppendEntries(args *AppendEntriesArgs, reply *AppendEntriesReply
 	}
 
 	if len(args.Entries) == 0 {
-		// TODO(heng): dealing with heatbeat here
+		// Handle heartbeat
```
Comment (liana-yang) [inline lab/src/raft/append_entries.go:43]: We should check `IsLogConsecutive` after the heartbeat.

File: lab/src/raft/leader_election.go [hunk 3]
```diff
@@ -1,11 +1,32 @@
...
+			reply := AppendEntriesReply{}
+			ok := rf.SendAppendEntries(server, &args, &reply)
+			if !ok {
+				return
+			}
+			if !reply.Success && reply.Term > currentTerm {
```
Comment (liana-yang) [inline lab/src/raft/leader_election.go]: We only have to check `reply.Succes` here.

File: lab/src/raft/leader_election.go (hunk 1 above)
Comment (flyingmeat) [reply lab/src/raft/leader_election.go:19]: Oh Actually leaderId is already in the AppendEntriesArgs. I did not check this carefully :\

File: lab/src/raft/leader_election.go (hunk 3 above)
Comment (flyingmeat) [reply lab/src/raft/leader_election.go]: Done

File: lab/src/raft/append_entries.go (hunk 2 above)
Comment (flyingmeat) [reply lab/src/raft/append_entries.go:43]: Done

**Comment by flyingmeat** [comment]
PING @Shauro

**Review by liana-yang — APPROVED** [review]

**Pull request merged**